rp2040-hal/src/multicore.rs

```diff
@@ -140,13 +140,36 @@ impl<'p> Core<'p> {
         }
     }
 
-    fn inner_spawn(
-        &mut self,
-        wrapper: *mut (),
-        entry: *mut (),
-        stack: &'static mut [usize],
-    ) -> Result<(), Error> {
-        if let Some((psm, ppb, sio)) = self.inner.as_mut() {
+    /// Spawn a function on this core.
+    pub fn spawn<F>(&mut self, stack: &'static mut [usize], entry: F) -> Result<(), Error>
+    where
+        F: FnOnce() -> bad::Never + Send + 'static,
+    {
+        if let Some((psm, ppb, fifo)) = self.inner.as_mut() {
+            // The first two ignored `u64` parameters are there to take up all of the registers,
+            // which means that the rest of the arguments are taken from the stack,
+            // where we're able to put them from core 0.
+            extern "C" fn core1_startup<F: FnOnce() -> bad::Never>(
+                _: u64,
+                _: u64,
+                entry: &mut ManuallyDrop<F>,
```

Comment:
I wondered if it was ok to have `&mut` argument, but write `*mut` pointer to the stack. Even if it's an obvious choice to have the same memory representation for `*` and `&`, it's not necessarily guaranteed.

After doing some research I came to the conclusion that this is actually fine: https://rust-lang.github.io/unsafe-code-guidelines/layout/pointers.html says: "The layouts of &T, &mut T, *const T and *mut T are the same."

Is this common knowledge (of people reading low-level code), or should it be explained in a comment?

Reply:
I'm actually a little more concerned about how it's casted to a `usize` before writing and read back as a pointer - that seems like it's _probably_ equivalent to a `mem::transmute`, and under [strict provenance](https://doc.rust-lang.org/nightly/std/ptr/index.html#strict-provenance), [apparently that produces an invalid pointer](https://www.reddit.com/r/rust/comments/u1bbqn/pointers_are_complicated_iii_or_pointerinteger/iaarpcg/).

Reply:
To be safe, I've changed the stack setup to write the pointers directly rather than casting to `usize`.

Reply:
No, casting is absolutely *not* equivalent to transmute. The reddit comment by Ralf Jung you linked specifically states that pointer-int casts and int-pointer casts are something else entirely, and the rest of the thread discusses this several times.
Using the correct types for these doesn't cost us anything (and it's better to preserve semantic information than to discard it), but also keep in mind we're spending time (both changing the code and discussing it) to conform to memory semantics Rust does not now and might never have.

Reply:
> No, casting is absolutely *not* equivalent to transmute.

Sorry, I was unclear - I know that `as` casts aren't the same as transmutes, I meant that it seemed like a transmute for the function to reinterpret the `usize` written to the stack as an `&mut ...`.

> but also keep in mind we're spending time (both changing the code and discussing it) to conform to memory semantics Rust does not now and might never have.

Fair enough.